src/main/java/com/ibm/disni/util/MemoryAllocation.java

```diff
@@ -21,67 +21,155 @@
-		} 
+
+	private int roundUpSize(int size){
+		// Round up length to the nearest power of two, or the minimum block size
+		if (size < MIN_BLOCK_SIZE) {
+			size = MIN_BLOCK_SIZE;
+		} else {
+			size--;
+			size |= size >> 1;
+			size |= size >> 2;
+			size |= size >> 4;
+			size |= size >> 8;
+			size |= size >> 16;
+			size++;
+		}
+		return size;
 	}
 
-	private String generateKey(int size, MemType type, String classname) {
-		String key = classname + ":" + size + ":" + type.ordinal();
-		return key;
+	public MemBuf allocate(int size){
+		return getOrCreateAllocatorStack(roundUpSize(size)).get();
 	}
 
-	private MemBuf _allocate(int size, MemType type, String classname) {
-		if (type == MemType.INDIRECT) {
-			ByteBuffer buffer = ByteBuffer.allocate(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = 0;
-			return new MemBuf(MemType.INDIRECT, buffer, address, this,
-					classname);
-		} else {
-			ByteBuffer buffer = ByteBuffer.allocateDirect(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = MemoryUtils.getAddress(buffer);
-			return new MemBuf(MemType.DIRECT, buffer, address, this, classname);
+	private AllocatorStack getOrCreateAllocatorStack(int length) {
+		AllocatorStack allocatorStack = allocStackMap.get(length);
+		if (allocatorStack == null) {
+			allocStackMap.putIfAbsent(length, new AllocatorStack(length));
+			allocatorStack = allocStackMap.get(length);
+		}
+
+		return allocatorStack;
+	}
+
+	public void put(MemBuf buf){
+		AllocatorStack allocatorStack = allocStackMap.get(buf.size());
+		if (allocatorStack != null) {
+			allocatorStack.put(buf);
+			if (startedCleanStacks.compareAndSet(false, true)) {
+				// Check the size of current idling buffers
+				long idleBuffersSize = allocStackMap
+					    .reduceValuesToLong(100L,
+							  allocStack -> allocStack.idleBuffersSize.get(), 0L, Long::sum);
+				// If it reached out 90% of idle buffer capacity, clean old stacks
+				if (idleBuffersSize > MAX_CACHE_SIZE * 0.90) {
+					cleanLRUSBuffers(idleBuffersSize);
+				} else {
+					startedCleanStacks.compareAndSet(true, false);
+				}
+			}
 		}
 	}
+
+	private void cleanLRUSBuffers(long idleBuffersSize){
+		logger.debug("Current idling buffers size {}KB exceed 90% of maxCacheSize {}KB." +
+			" Cleaning LRU idling buffers", idleBuffersSize / 1024, MAX_CACHE_SIZE / 1024);
+		// Find least recently used buffer stack - it has lowest lastAccess value
+		AllocatorStack lruStack = allocStackMap.values().stream()
+			.filter(s -> !s.stack.isEmpty())
+			.sorted(Comparator.comparingLong(s -> s.lastAccess)).iterator().next();
+		long totalCleaned = 0;
+		// Will clean up to 65% of capacity
+		long needToClean = idleBuffersSize - (long) (MAX_CACHE_SIZE * 0.65);
+		while (!lruStack.stack.isEmpty() && totalCleaned < needToClean) {
+			MemBuf buffer = lruStack.stack.pollFirst();
+			if (buffer != null) {
+				totalCleaned += lruStack.size;
+				lruStack.idleBuffersSize.addAndGet(-lruStack.size);
```

Comment:
Shouldn't this be buffer size instead of stack size?

Reply:
Each buffer in the stack is of the same size.

Reply:
Right, makes sense.